model/labels/labels.go

```diff
@@ -281,68 +290,83 @@ func (ls Labels) Has(name string) bool {
 // HasDuplicateLabelNames returns whether ls has duplicate label names.
 // It assumes that the labelset is sorted.
 func (ls Labels) HasDuplicateLabelNames() (string, bool) {
-	for i, l := range ls {
-		if i == 0 {
-			continue
-		}
-		if l.Name == ls[i-1].Name {
-			return l.Name, true
+	var lName, prevName string
+	for i := 0; i < len(ls.data); {
+		lName, i = decodeString(ls.data, i)
+		_, i = decodeString(ls.data, i)
+		if lName == prevName {
+			return lName, true
 		}
+		prevName = lName
 	}
 	return "", false
 }
 
 // WithoutEmpty returns the labelset without empty labels.
 // May return the same labelset.
 func (ls Labels) WithoutEmpty() Labels {
-	for _, v := range ls {
-		if v.Value != "" {
+	for pos := 0; pos < len(ls.data); {
+		_, newPos := decodeString(ls.data, pos)
+		lValue, newPos := decodeString(ls.data, newPos)
+		if lValue != "" {
+			pos = newPos
 			continue
 		}
 		// Do not copy the slice until it's necessary.
-		els := make(Labels, 0, len(ls)-1)
-		for _, v := range ls {
-			if v.Value != "" {
-				els = append(els, v)
+		// TODO: could optimise the case where all blanks are at the end.
+		// Note: we size the new buffer on the assumption there is exactly one blank value.
+		buf := make([]byte, pos, pos+(len(ls.data)-newPos))
+		copy(buf, ls.data[:pos]) // copy the initial non-blank labels
+		pos = newPos             // move past the first blank value
+		for pos < len(ls.data) {
+			var newPos int
+			_, newPos = decodeString(ls.data, pos)
+			lValue, newPos = decodeString(ls.data, newPos)
+			if lValue != "" {
+				buf = append(buf, ls.data[pos:newPos]...)
 			}
+			pos = newPos
 		}
-		return els
+		return Labels{data: yoloString(buf)}
 	}
 	return ls
 }
 
 // Equal returns whether the two label sets are equal.
 func Equal(ls, o Labels) bool {
-	if len(ls) != len(o) {
-		return false
-	}
-	for i, l := range ls {
-		if l != o[i] {
-			return false
-		}
-	}
-	return true
+	return ls.data == o.data
 }
 
 // Map returns a string map of the labels.
 func (ls Labels) Map() map[string]string {
-	m := make(map[string]string, len(ls))
-	for _, l := range ls {
-		m[l.Name] = l.Value
+	m := make(map[string]string, len(ls.data)/10)
```

Comment:
Where does that 10 come from?
From my understanding so far it should be 2+something for the length information.

Reply:
This is to size the map, which will contain all label names and values.  It would be correct if all label names and values are length 4; the map size will be a bit of an over-estimate in most cases, which is ok.  Probably deserving a comment.